### PR TITLE
Protect Indexed_pool.add_from_gossip_exn from changes to the ledger

### LINKED ISSUE #4255: 1dffa7c: [Fatal] Unhandled top-level exception: $exn
```
$ coda daemon -discovery-port 8303    
-peer /dns4/seed-one.genesis.o1test.net/tcp/10002/ipfs/12D3KooWP7fTKbyiUcYJGajQDpCFo2rDexgTHFJTxCH8jvcL1eAH    
-peer /dns4/seed-two.genesis.o1test.net/tcp/10002/ipfs/12D3KooWL9ywbiXNfMBqnUKHSB1Q1BaHFNUzppu6JLMVn9TTPFSA 
-run-snark-worker $CODA_PK    
-snark-worker-fee 1 
-propose-key keys/my-wallet
```
Daemon bootstrapped and synced, after few hours crashed
```
[Fatal] Unhandled top-level exception: $exn
Generating crash report
        exn: "(monitor.ml.Error \"Assert_failure src/lib/network_pool/indexed_pool.ml:529:8\"\n  
(\"Raised at file \\\"src/lib/network_pool/indexed_pool.ml\\\", line 529, characters 8-69\"\n    
\"Called from file \\\"src/lib/network_pool/transaction_pool.ml\\\", line 531, characters 30-193\"\n 
   \"Called from file \\\"src/lib/network_pool/network_pool_base.ml\\\", line 26, characters 15-65\"\n 
   \"Called from file \\\"src/pipe.ml\\\", line 827, characters 4-56\"\n
    \"Called from file \\\"src/job_queue.ml\\\" (inlined), line
131, characters 2-5\"\n
    \"Called from file \\\"src/job_queue.ml\\\", line 171, characters 6-47\"\n
    \"Caught by monitor coda\"))"
```
```
$ coda version
Commit 1dffa7c on branch master
```
[coda_crash_report_2020-01-25_07-47-18.817861.tar.gz](https://github.com/CodaProtocol/coda/files/4111926/coda_crash_report_2020-01-25_07-47-18.817861.tar.gz)

### LINKED ISSUE #4274: Unhandled exception: network client EOF
👋coda! Experienced this crash earlier while running testnet node: 
```
error: "(src/connection.ml.Handshake_error.Handshake_error\n (Eof (\"Client connected via TCP\" (165.22.241.104 8302))))"
2020-01-25 22:56:25 UTC [Error] RPC call raised an exception: $exn
        exn: "(src/connection.ml.Handshake_error.Handshake_error\n  (Eof (\"Client connected via TCP\" (87.2.108.13 8302))))"
2020-01-25 22:56:25 UTC [Error] Broadcasting message "transaction pool diff" to $peer failed: $error
        peer: { "host": "87.2.108.13", "discovery_port": 8303, "communication_port": 8302 }
        error: "(src/connection.ml.Handshake_error.Handshake_error\n (Eof (\"Client connected via TCP\" (87.2.108.13 8302))))"
2020-01-25 22:57:13 UTC [Fatal] Unhandled top-level exception: $exn
Generating crash report
        exn: "(monitor.ml.Error \"Assert_failure src/lib/network_pool/indexed_pool.ml:529:8\"\n  (\"Raised at file \\\"src/lib/network_pool/indexed_pool.ml\\\", line 529, characters 8-69\"\n    \"Called from file \\\"src/lib/network_pool/transaction_pool.ml\\\", line 531, characters 30-193\"\n    \"Called from file \\\"src/lib/network_pool/network_pool_base.ml\\\", line 26, characters 15-65\"\n    \"Called from file \\\"src/pipe.ml\\\", line 827, characters 4-56\"\n    \"Called from file \\\"src/job_queue.ml\\\" (inlined), line 131, characters 2-5\"\n    \"Called from file \\\"src/job_queue.ml\\\", line 171, characters 6-47\"\n    \"Caught by monitor coda\"))"
```
Crash report: [coda_crash_report_2020-01-25_22-57-13.552780.tar.gz](https://github.com/CodaProtocol/coda/files/4113052/coda_crash_report_2020-01-25_22-57-13.552780.tar.gz)

### LINKED ISSUE #4282: Coda Daemon crashed: Assert_failure src/lib/network_pool/indexed_pool.ml:529:8
"(monitor.ml.Error \"Assert_failure src/lib/network_pool/indexed_pool.ml:529:8\"\n  (\"Raised at file \\\"src/lib/network_pool/indexed_pool.ml\\\", line 529, characters 8-69\"\n    \"Called from file \\\"src/lib/network_pool/transaction_pool.ml\\\", line 531, characters 30-193\"\n    \"Called from file \\\"src/lib/network_pool/network_pool_base.ml\\\", line 26, characters 15-65\"\n    \"Called from file \\\"src/pipe.ml\\\", line 827, characters 4-56\"\n    \"Called from file \\\"src/job_queue.ml\\\" (inlined), line 131, characters 2-5\"\n    \"Called from file \\\"src/job_queue.ml\\\", line 171, characters 6-47\"\n    \"Caught by monitor coda\"))"

### LINKED ISSUE #7613: Transaction pool may attempt to add transactions between ledger update and pool revalidation
[This assertion](https://github.com/MinaProtocol/mina/blob/cb0ed008f1d39dba838f3dcc5bcf5ca1f0e30362/src/lib/network_pool/indexed_pool.ml#L829) catches this case, and was reported as happening on our nodes in a testnet in #6951, as well as previously in #4255, #4274,  #4282.

My hypothesis is that
* the ledger had updated, including a change to a particular account, but `Indexed_pool.revalidate` had not yet been run
* that account had transactions pending in the pool
* another transaction for that account was gossiped and was being validated

Suggested fix is to
* store the root hash of the ledger in the indexed pool at every revalidation
* before attempting to add a batch of gossiped transactions, check that the ledger hash matches the one that the pool is up-to-date with
  - if it isn't, run `Indexed_pool.wait_for_next_revalidation : unit -> Ledger_hash.t Deferred.t` (not yet implemented)
* before adding *each transaction*, check that the ledger hash hasn't changed
  - the ledger can potentially change out from underneath us because we're sharing a mutable ledger among several threads :weary:
  - if it has, interrupt the current evaluation (since we have modified our copy of the pool, and revalidation may also have modified it)
  - either restart the evaluation (better option) or throw away the transaction pool diff (worse, easier option)

This is a subtle variation on #6946.

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR interrupts adding transactions when there is inconsistent state between a ledger and the transaction pool. This avoids an assertion error caused by an invariant that this breaks. #7613 contains a fuller write-up of what is happening here.

Checklist:

- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [X] Does this close issues? List them:

Closes #7613
Closes #4255
Closes #4274
Closes #4282

### CODE REVIEW

File: src/lib/network_pool/transaction_pool.ml [hunk 1]
```diff
@@ -926,260 +959,290 @@ struct
...
-                                      , [ ("cmd", User_command.to_yojson tx)
-                                        ; ("reason", yojson_fail_reason err)
-                                        ; ("error_extra", `Assoc error_extra)
-                                        ] ) )
-                              in
-                              go txs'' pool
-                                (accepted, (tx, diff_err) :: rejected) )
-                        else
+                    let new_ledger_hash = Base_ledger.merkle_root ledger in
+                    if not (Ledger_hash.equal ledger_hash new_ledger_hash) then
```
Comment (deepthiskumar) [inline src/lib/network_pool/transaction_pool.ml:963]: When is the ledger mask updated? every time the ledger changes, best_tip_ledger gets a new mask. `ledger` here would only fail with dangling parent, no? Did you mean to get the updated ledger from `best_tip_ledger` again?

Comment (mrmr1993) [reply src/lib/network_pool/transaction_pool.ml:963]: This PR is in response to #7613, where the hypothesis is that the ledger is being mutated. This is to protect against that mutation.

File: src/lib/network_pool/transaction_pool.ml [hunk 2]
```diff
@@ -897,6 +903,33 @@ struct
...
+                    | Some ledger
+                      when Ledger_hash.equal new_ledger_hash
+                             (Base_ledger.merkle_root ledger) ->
+                        Deferred.Or_error.return ledger
+                    | _ ->
+                        Deferred.Or_error.error_string
+                          "Could not find the ledger against which the \
+                           transaction pool is valid."
+                in
+                (ledger_hash, ledger)
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/lib/network_pool/transaction_pool.ml:930]: We've just checked that it's equal above, and errored out if it isn't.

**Comment by mrmr1993** [comment]
We haven't seen the error from #7613 again since #7759 was merged. Hopefully, this is no longer needed. Closing.

**Pull request closed without merging**